src/lib/SaferNames/NameCore.hs

```diff
@@ -0,0 +1,294 @@
+
+extendNameSet :: NameSet n -> NameSet (n:=>:l) -> NameSet l
+extendNameSet (UnsafeMakeNameSet s1) (UnsafeMakeNameSet s2) =
+  UnsafeMakeNameSet (s1 <> s2)
+
+concatNameSets :: NameSet (n1:=>:n2) -> NameSet (n2:=>:n3) -> NameSet (n1:=>:n3)
+concatNameSets (UnsafeMakeNameSet s1) (UnsafeMakeNameSet s2) =
+  UnsafeMakeNameSet (s1 <> s2)
+
+-- TODO: we reuse the old `Name` to make use of the GlobalName name space while
+-- we're using both the old and new systems together.
+-- TODO: something like this instead:
+--    type Tag = T.Text
+--    data RawName = RawName Tag Int deriving (Show, Eq, Ord)
+type RawName = D.Name
+
+data Name (s::E)  -- static information associated with name
+          (n::S)  -- scope parameter
+  where
+    UnsafeMakeName :: Typeable s => RawName -> Name s n
+
+data NameBinder (s::E)  -- static information for the name this binds (note
+                         -- that `NameBinder` doesn't actually carry this data)
+                 (n::S)  -- scope the binder lives in
+                 (l::S)  -- scope within the binder's scope
+  where
+    UnsafeMakeBinder :: Typeable s => RawName -> NameBinder s n l
+    Ignore           :: NameBinder s n n
+
+withFresh :: Typeable s => Distinct n => NameSet n
+         -> (forall l. Distinct l => NameBinder s n l -> Name s l -> a) -> a
+withFresh (UnsafeMakeNameSet scope) cont =
+  cont @UnsafeMakeDistinctS (UnsafeMakeBinder freshName) (UnsafeMakeName freshName)
+  where
+    freshName = freshRawName "v" scope
+
+freshRawName :: D.Tag -> S.Set RawName -> RawName
+freshRawName tag usedNames = D.Name D.GenName tag nextNum
+  where
+    nextNum = case S.lookupLT (D.Name D.GenName tag bigInt) usedNames of
+                Just (D.Name D.GenName tag' i)
+                  | tag' /= tag -> 0
+                  | i < bigInt  -> i + 1
+                  | otherwise   -> error "Ran out of numbers!"
+                _ -> 0
+    bigInt = (10::Int) ^ (9::Int)  -- TODO: consider a real sentinel value
+
+projectName :: NameSet (n:=>:l) -> Name s l -> Either (Name s n) (Name s (n:=>:l))
+projectName (UnsafeMakeNameSet scope) (UnsafeMakeName rawName)
+  | S.member rawName scope = Right $ UnsafeMakeName rawName
+  | otherwise              = Left  $ UnsafeMakeName rawName
+
+-- proves that the names in n are distinct
+class Distinct (n::S)
+
+instance Distinct VoidS
+instance Distinct UnsafeMakeDistinctS
+
+-- === injections ===
+
+-- When we inline an expression, we lift it into a larger (deeper) scope,
+-- containing more in-scope variables. For example, when we turn this:
+--   let foo = \x. x + z
+--   in \y. foo y
+-- into this:
+--   \y. (\x. x + z) y
+-- The expression `\x. x + z`, initially in the scope `[z]`, gets injected into
+-- the scope `[z, y]`. To prove that we can do the injection as an coercion,
+-- it's sufficient to know (1) that the new scope contains distinct names, and
+-- (2) it extends the old scope. These are the `Distinct l` and
+-- `(InjectableB b, b)` conditions below.
```

Comment:
Exactly. It's sufficient but not necessary. Until yesterday, I'd been working with the weaker condition I'd called `FreshExt`, which says that you can inject a name from `n` to `l` if you know that `l` extends `n` with fresh names. But it turned out that in practice, we always met the stronger condition that the names in `l` were unique in an absolute sense, not just a relative one. And the single-parameter `Distinct n` is a bit easier to work with than the two-parameter `FreshExt n l`, so I switched to that.

Reply:
Actually, I'm still a bit confused about this. Why do we need `InjectableE` or `InjectableB` in the definition below? Isn't it enough to just know that every name is distinct, and that there exists some binder witnessing the extension?

Reply:
Yes, it's stale. `(InjectableB, b)` became `NameSet (n:=>:l)`. @danieldjohnson the `InjectableE` condition is still necessary because not everything with kind `S -> *` can be injected. It's the usual contravariance issue with subtyping. A map from names `Name n` isn't a valid map if you extend `n` to `l`.

Reply:
Hm, ok. Should that be a third condition, then? Or is the issue that this comment is only talking about expressions, but the implementation is talking about any `E`? Maybe an example of an `E` that isn't `InjectableE` would be useful